File: 2024/day06_guard_gallivant/main.py
import copy
import argparse


def read_input(filename):
    raw_input = []
    start_pos = (0, 0)
    with open(filename, 'r') as file:
        for line in file:
            pos = line.find('^')
            if pos != -1:
                start_pos = (len(raw_input), pos)
            raw_input.append(line)

    room = [list(line.strip()) for line in raw_input]
    return room, start_pos


direction = {
    '^': (-1, 0),
    '>': (0,  1),
    'v': (1,  0),
    '<': (0, -1)
}


def turn_right(dir):
    match dir:
        case '^': return '>'
        case '>': return 'v'
        case 'v': return '<'
        case '<': return '^'


def next_step(cur_pos, dir):
    return (cur_pos[0] + direction[dir][0], cur_pos[1] + direction[dir][1])


def patrol(room, start_pos, obstacle=None):
    h = len(room)-1
    w = len(room[0])-1
    dir = '^'
    cur_pos = start_pos
    room[start_pos[0]][start_pos[1]] = 'X'
    dir_change = []
    if obstacle is not None:
        room[obstacle[0]][obstacle[1]] = 'O'
    while True:
        room[cur_pos[0]][cur_pos[1]] = 'X'
        next_pos = next_step(cur_pos, dir)
        if next_pos[0] < 0 or next_pos[0] > w or next_pos[1] < 0 or next_pos[1] > h:
            break
        if (room[next_pos[0]][next_pos[1]] == '#' or room[next_pos[0]][next_pos[1]] == 'O'):
            bump = (cur_pos[0], cur_pos[1], dir)
            if bump in dir_change:
                return -1  # found a cycle
            dir_change.append(bump)
            dir = turn_right(dir)
        else:
            cur_pos = next_pos

    return sum([line.count('X') for line in room]), room


def find_patrol_cycles(room, start_pos):
    h = len(room) - 1
    w = len(room[0]) - 1

    _, patrolled_room = patrol(room, start_pos)

    obstacles_that_cause_cycle = []
    for r in range(h):
        for c in range(w):
            if patrolled_room[r][c] != 'X':
                continue
            s = patrol(copy.deepcopy(room), start_pos, (r, c))
            if s == -1:
                obstacles_that_cause_cycle.append((r, c))

    return len(obstacles_that_cause_cycle)


if __name__ == '__main__':
    parser = argparse.ArgumentParser("main.py")
    parser.add_argument("filename", nargs='?',
                        help="input filename", default="input.txt")
    args = parser.parse_args()

    room, start_pos = read_input(args.filename)

    steps_covered, _ = patrol(room, start_pos)
    num_cycle_obstacles = find_patrol_cycles(room, start_pos)

    print(f"\nPart1: steps_covered = {steps_covered}")
    print(f"Part2: num_cycle_obstacles = {num_cycle_obstacles}\n")

File: 2024/day06_guard_gallivant/main.rs
use std::collections::{BTreeMap, BTreeSet};
use std::fs;
use std::time::Instant;

#[derive(Debug)]
struct Input {
    row_obstacles: BTreeMap<usize, Vec<usize>>,
    col_obstacles: BTreeMap<usize, Vec<usize>>,
    start_pos: (usize, usize),
    num_rows: usize,
    num_cols: usize,
}

fn read_input(filename: &str) -> Input {
    let mut row_obstacles: BTreeMap<usize, Vec<usize>> = BTreeMap::new();
    let mut col_obstacles: BTreeMap<usize, Vec<usize>> = BTreeMap::new();
    let mut start_pos = None;
    let mut num_rows = 0;
    let mut num_cols = 0;

    let content = fs::read_to_string(filename).expect("Failed to read file");

    for (r, line) in content.lines().enumerate() {
        if start_pos.is_none() {
            if let Some(pos) = line.find('^') {
                start_pos = Some((r, pos));
            }
        }

        let obs: Vec<usize> = line
            .chars()
            .enumerate()
            .filter_map(|(c, ch)| if ch == '#' { Some(c) } else { None })
            .collect();

        if !obs.is_empty() {
            row_obstacles.insert(r, obs.clone());
            for &c in &obs {
                col_obstacles.entry(c).or_insert_with(Vec::new).push(r);
            }
        }

        num_rows += 1;
        num_cols = line.len();
    }

    Input {
        row_obstacles,
        col_obstacles,
        start_pos: start_pos.expect("No starting position found"),
        num_rows,
        num_cols,
    }
}

fn first_less_than(items: &[usize], x: usize) -> Option<usize> {
    items.iter().rev().find(|&&item| item < x).cloned()
}

fn first_greater_than(items: &[usize], x: usize) -> Option<usize> {
    items.iter().find(|&&item| item > x).cloned()
}

fn patrol(
    row_obstacles: &mut BTreeMap<usize, Vec<usize>>,
    col_obstacles: &mut BTreeMap<usize, Vec<usize>>,
    num_rows: usize,
    num_cols: usize,
    start_pos: (usize, usize),
) -> Option<BTreeSet<(usize, usize, usize, usize)>> {
    let mut dir = '^';
    let (mut r, mut c) = start_pos;
    let mut segments = BTreeSet::new();

    loop {
        let (mut new_r, mut new_c) = (r, c);
        match dir {
            '^' => {
                if let Some(obs_r) = col_obstacles
                    .get(&c)
                    .and_then(|obs| first_less_than(obs, r))
                {
                    new_r = obs_r + 1;
                    dir = '>';
                } else {
                    segments.insert((r, c, 0, c));
                    break;
                }
            }
            '>' => {
                if let Some(obs_c) = row_obstacles
                    .get(&r)
                    .and_then(|obs| first_greater_than(obs, c))
                {
                    new_c = obs_c - 1;
                    dir = 'v';
                } else {
                    segments.insert((r, c, r, num_cols - 1));
                    break;
                }
            }
            'v' => {
                if let Some(obs_r) = col_obstacles
                    .get(&c)
                    .and_then(|obs| first_greater_than(obs, r))
                {
                    new_r = obs_r - 1;
                    dir = '<';
                } else {
                    segments.insert((r, c, num_rows - 1, c));
                    break;
                }
            }
            '<' => {
                if let Some(obs_c) = row_obstacles
                    .get(&r)
                    .and_then(|obs| first_less_than(obs, c))
                {
                    new_c = obs_c + 1;
                    dir = '^';
                } else {
                    segments.insert((r, c, r, 0));
                    break;
                }
            }
            _ => unreachable!(),
        }

        let new_segment = (r, c, new_r, new_c);
        if segments.contains(&new_segment) {
            return None;
        }
        segments.insert(new_segment);
        r = new_r;
        c = new_c;
    }

    Some(segments)
}

fn get_covered_points(
    segments: &BTreeSet<(usize, usize, usize, usize)>,
) -> BTreeSet<(usize, usize)> {
    let mut covered_points = BTreeSet::new();

    for &(r1, c1, r2, c2) in segments {
        if r1 == r2 {
            for c in c1.min(c2)..=c1.max(c2) {
                covered_points.insert((r1, c));
            }
        } else if c1 == c2 {
            for r in r1.min(r2)..=r1.max(r2) {
                covered_points.insert((r, c1));
            }
        }
    }

    covered_points
}

fn find_patrol_cycles(
    covered_points: &BTreeSet<(usize, usize)>,
    row_obstacles: &mut BTreeMap<usize, Vec<usize>>,
    col_obstacles: &mut BTreeMap<usize, Vec<usize>>,
    num_rows: usize,
    num_cols: usize,
    start_pos: (usize, usize),
) -> usize {
    let mut num_cycles = 0;

    for &(r, c) in covered_points {
        row_obstacles.entry(r).or_default().push(c);
        row_obstacles.get_mut(&r).unwrap().sort();

        col_obstacles.entry(c).or_default().push(r);
        col_obstacles.get_mut(&c).unwrap().sort();

        if patrol(row_obstacles, col_obstacles, num_rows, num_cols, start_pos).is_none() {
            num_cycles += 1;
        }

        row_obstacles.get_mut(&r).unwrap().retain(|&x| x != c);
        col_obstacles.get_mut(&c).unwrap().retain(|&x| x != r);
    }

    num_cycles
}

fn main() {
    let filename = "input.txt";
    let input = read_input(filename);

    // Part 1
    let part1_start = Instant::now();

    let segments = patrol(
        &mut input.row_obstacles.clone(),
        &mut input.col_obstacles.clone(),
        input.num_rows,
        input.num_cols,
        input.start_pos,
    )
    .expect("No segments found");

    let covered_points = get_covered_points(&segments);
    let part1_duration = part1_start.elapsed();

    println!(
        "Part1: steps_covered = {} (Elapsed: {:.2?})",
        covered_points.len(),
        part1_duration
    );

    // Part 2
    let part2_start = Instant::now();

    let num_cycles = find_patrol_cycles(
        &covered_points,
        &mut input.row_obstacles.clone(),
        &mut input.col_obstacles.clone(),
        input.num_rows,
        input.num_cols,
        input.start_pos,
    );
    let part2_duration = part2_start.elapsed();

    println!(
        "Part2: num_cycles = {} (Elapsed: {:.2?})",
        num_cycles, part2_duration
    );
}
